driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java

```diff
@@ -215,13 +218,17 @@ static void encodeMessage(final RequestMessage message, final BsonOutput bsonOut
         }
     }
 
+    private static final List<Integer> NOT_MASTER_CODES = asList(10107, 13435);
+    private static final List<Integer> RECOVERING_CODES = asList(11600, 11602, 13436, 189, 91);
     private static MongoException createSpecialException(final BsonDocument response, final ServerAddress serverAddress,
```

Comment:
The spec also says:

> Additionally, if the response includes a write concern error, then the code and message of the write concern error MUST be checked the same way a response error is checked above.

Since write concern is checked in the operation layer, it's going to require a bit more work to detect write concern errors at this level in order to invalidate the server/pool but without actually throwing the exception.

Reply:
Updated the command transformers and isRetryable helper

Reply:
By 

> MUST be checked the same way a response error is checked above

I think the spec is referring to text that's further down in the same section:

> ...it MUST replace the server's description with a default ServerDescription of type Unknown. It MUST store useful information in the new ServerDescription's error field, including the error message from the server.

Currently that behavior is implemented in DefaultServer, but in order to do the same for this class of write concern errors, a write concern exception would have to be detected and perhaps thrown at a much lower level than it is now, probably down in InternalStreamConnection.  I'm not sure though what the ramifications of doing that would be.

Reply:
@jyemin I'm not clear on the issue here. This PR checks the response already to see if it contains a write concern error in the ProtocolHelper.createSpecialException - it would then in turn trigger the invalidation of the pool. Are you saying write concern commands wouldnt fall through this code path?